src/azure-cli-core/azure/cli/core/parser.py

```diff
@@ -280,37 +280,157 @@ def parse_known_args(self, args=None, namespace=None):
+                return None
+            if response.status_code == 200:
+                EXT_CMD_INDEX.data = response.json()
+                EXT_CMD_INDEX.save_with_retry()
+            else:
+                logger.info("Error when retrieving extension command index. Response code: %s", response.status_code)
+                return None
+        cmd_chain = EXT_CMD_INDEX
+        for part in command_str.split():
+            try:
+                if isinstance(cmd_chain[part], str):
+                    return cmd_chain[part]
+                cmd_chain = cmd_chain[part]
+            except KeyError:
+                return None
+        return None
+
+    def _check_value(self, action, value):  # pylint: disable=too-many-statements
         # Override to customize the error message when a argument is not among the available choices
         # converted value must be one of the choices (if specified)
-        if action.choices is not None and value not in action.choices:
+        if action.choices is not None and value not in action.choices:  # pylint: disable=too-many-nested-blocks
+            caused_by_extension_not_installed = False
             if not self.command_source:
-                # parser has no `command_source`, value is part of command itself
-                extensions_link = 'https://docs.microsoft.com/en-us/cli/azure/azure-cli-extensions-overview'
-                error_msg = ("{prog}: '{value}' is not in the '{prog}' command group. See '{prog} --help'. "
-                             "If the command is from an extension, "
-                             "please make sure the corresponding extension is installed. "
-                             "To learn more about extensions, please visit "
-                             "{extensions_link}").format(prog=self.prog, value=value, extensions_link=extensions_link)
+                candidates = difflib.get_close_matches(value, action.choices, cutoff=0.7)
+                error_msg = None
+                # self.cli_ctx is None when self.prog is beyond 'az', such as 'az iot'.
+                # use cli_ctx from cli_help which is not lost.
+                cli_ctx = self.cli_ctx or (self.cli_help.cli_ctx if self.cli_help else None)
+                use_dynamic_install = cli_ctx.config.get(
+                    'extension', 'use_dynamic_install', 'no') if cli_ctx else 'no'
+                if use_dynamic_install.lower() not in ['no', 'yes_prompt', 'yes_without_prompt']:
+                    use_dynamic_install = 'no'
+                if use_dynamic_install.lower() != 'no' and not candidates:
+                    # Check if the command is from an extension
+                    from azure.cli.core.util import roughly_parse_command
+                    cmd_list = self.prog.split() + self._raw_arguments
+                    command_str = roughly_parse_command(cmd_list[1:])
+                    ext_name = self._search_in_extension_commands(command_str)
+                    if ext_name:
```

Comment:
what if `ext_name` is `None` ?

Reply:
If `ext_name` is `None`, `error_msg` will be `None` and below code will use the `{prog}: '{value}' is not in the '{prog}' command group.` for error message.